firestore-db-integration

balzhan-kanatbek opened the pull request.

balzhan-kanatbek committed:
feat: enhance authentication flow in App component

- Added Sign Up and Login pages to the navigation.
- Integrated authentication context to manage user login state.
- Implemented logout functionality to reset the current page to home.
- Updated rendering logic to conditionally display ProfilePage based on authentication status.

balzhan-kanatbek committed:
feat: implement authentication context and user interface for login and signup

balzhan-kanatbek committed:
feat: update authentication and quote management features

- Added Firebase integration for user authentication.
- Refactored AuthContextProvider to handle login, signup, and Google sign-in.
- Introduced MyCollectionPage for users to view their liked and created quotes.
- Updated MainPage and QuoteCard components to conditionally render based on authentication status.
- Removed ProfilePage and adjusted routing logic accordingly.
- Enhanced user experience with password visibility toggle and improved error handling in login/signup forms.
- Updated package.json and .gitignore to include necessary configurations.

balzhan-kanatbek committed:
fix: update text for clarity on MyCollectionPage and SignUpPage

- Modified the description on MyCollectionPage to include "create" in the context of saving and viewing quotes.
- Changed the prompt on SignUpPage from "Sign In" to "Log In" for consistency in terminology.

balzhan-kanatbek committed:
Resolve merge conflicts

balzhan-kanatbek committed:
feat: add ProfilePage and update navigation logic

- Introduced ProfilePage component for authenticated users.
- Updated App component to include profile navigation and handle page changes using enum values for consistency.
- Adjusted button click handlers for login and signup to utilize enum for better maintainability.

balzhan-kanatbek committed:
feat: refactor App component and implement Navbar for improved navigation

- Replaced inline navigation logic in App component with a dedicated Navbar component for better separation of concerns.
- Updated routing logic to utilize React Router for navigation between pages, including protected routes for authenticated users.
- Refactored authentication handling in AuthContextProvider to streamline user state management.
- Enhanced mobile responsiveness and user experience in the Navbar with toggle functionality.

balzhan-kanatbek committed:
feat: enhance quote management and user authentication features

- Added support for liking and unliking quotes, allowing users to interact with quotes more dynamically.
- Refactored QuotesContextProvider and quotesReducer to manage liked quotes and user interactions effectively.
- Introduced new components for handling authentication forms, improving user experience during login and signup processes.
- Updated MainPage and MyCollectionPage to display liked and created quotes, enhancing user engagement.
- Implemented reusable components for input fields and social sign-in buttons to streamline form handling.
- Added routing updates to ensure seamless navigation between pages, including user authentication checks.

balzhan-kanatbek committed:
feat: integrate Firestore for quote management and update types

- Added Firestore integration to manage quotes, including fetching and adding quotes in real-time.
- Updated Quote interface to include an optional id property for better data handling.
- Refactored CreateQuoteForm to simplify the onCreate prop, aligning with the new quote management logic.
- Adjusted MainPage to utilize the new quote management features, enhancing user interaction.

balzhan-kanatbek committed:
feat: enhance quote management and user interface

- Updated App component to handle loading state during authentication, improving user experience.
- Refactored AuthContextProvider to include loading state and manage user authentication more effectively.
- Introduced EditQuoteForm component for editing existing quotes, enhancing user interaction.
- Updated QuoteCard to support editing and deleting quotes, allowing users to manage their content.
- Improved CreateQuoteForm to ensure quotes are associated with the authenticated user.
- Adjusted various components for better code consistency and readability.

annasvst commented on src/AuthContextProvider.tsx:
```diff
@@ -0,0 +1,152 @@
+} from "firebase/auth";
+
+import { auth } from "./firebase";
+import React from "react";
```
It's recommended to use named imports (this way you only import what you actually use) instead of default imports as it is right now.

annasvst commented on src/AuthContextProvider.tsx:
```diff
@@ -0,0 +1,152 @@
+}
+
+type AuthAction =
+  | { type: "LOGIN"; payload: { email: string; uid: string } }
```
Action type can be an enum

annasvst commented on src/components/CreateQuoteForm/index.tsx:
```diff
@@ -1,37 +1,50 @@
+      onCancel();
     } else {
-      alert("Please enter both quote and author.");
+      alert("Please enter both quote and author and ensure you are logged in.");
```
This is a very broad error message. Instead, don't show this page if the user is not logged in. You already have a check for that in the MyCollectionPage but to make the code more strict you can pass uid from MyCollectionPage to CreateQuoteForm as a required prop - this way you will be able to remove the 'not logged in' part from error validation in this component.

Add input validation - it should not be possible to submit the form with missing values. And make sure that you show the error messages to the user as text below inputs instead of alert.

annasvst commented on src/components/EditQuoteForm/index.tsx:
```diff
@@ -0,0 +1,92 @@
+      onSave();
+      onCancel();
+    } else {
+      alert("Please enter both quote and author.");
```
Again, add proper form validation instead of using alert

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,19 +1,86 @@
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 import { faThumbsUp as faThumbsUpSolid } from "@fortawesome/free-solid-svg-icons";
 import { faThumbsUp as faThumbsUpRegular } from "@fortawesome/free-regular-svg-icons";
 import { faThumbsDown as faThumbsDownRegular } from "@fortawesome/free-regular-svg-icons";
```
This import is not needed, remove

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,19 +1,86 @@
+  useQuotesContext,
+  useQuotesDispatch,
+} from "../../QuotesContextProvider";
 import { QuotesActionType } from "../../quotesReducer";
```
This import is not needed, remove

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,19 +1,86 @@
+  const { uid, isAuthenticated } = useAuth();
+  const { updateQuote, deleteQuote } = useFirestoreQuotes();
+  const { dispatch } = useQuotesDispatch();
   const { quotes } = useQuotesContext();
```
Not needed, remove

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,19 +1,86 @@
+}: QuoteCardProps) => {
+  const { uid, isAuthenticated } = useAuth();
+  const { updateQuote, deleteQuote } = useFirestoreQuotes();
+  const { dispatch } = useQuotesDispatch();
```
Not needed, remove

annasvst commented on src/pages/MainPage/index.tsx:
```diff
@@ -54,33 +49,15 @@ export const MainPage = () => {
-          onCancel={() => setShowCreateForm(false)}
+      {showQuote && currentQuote && (
+        <QuoteCard
+          quote={currentQuote.quote}
```
QuoteCard component expects id prop but it's not passed here

annasvst commented on src/hooks/useFirestoreQuotes.ts:
```diff
@@ -0,0 +1,75 @@
+import { db } from "../firebase";
+import { Quote } from "../types";
+
+export const useFirestoreQuotes = () => {
```
Well done here with creating a custom hook that encapsulated logic for CRUD firebase operations 👍
An improvement could be not to log success errors to console but instead return a message (success/error) so that it's possible to show a better confirmation to a user whenever this hook is used. You did similar think for auth form validation already.

balzhan-kanatbek committed:
feat: enhance quote forms and error handling

- Refactored CreateQuoteForm and EditQuoteForm to include validation for quote and author fields, improving user feedback.
- Updated AuthContextProvider to utilize enums for action types, enhancing code maintainability.
- Adjusted QuoteCard to streamline quote management and ensure proper user authentication.
- Enhanced useFirestoreQuotes hook to return operation results, providing better error handling during Firestore interactions.
- Updated MyCollectionPage to pass user ID to CreateQuoteForm, ensuring quotes are associated with the authenticated user.
